src/libraries/System.IO.Pipelines/src/System/IO/Pipelines/StreamPipeReader.cs

```diff
@@ -259,6 +259,142 @@ public override async ValueTask<ReadResult> ReadAsync(CancellationToken cancella
             }
         }
 
+        /// <inheritdoc />
+        public override async Task CopyToAsync(PipeWriter destination, CancellationToken cancellationToken = default)
+        {
+            ThrowIfCompleted();
+
+            // PERF: store InternalTokenSource locally to avoid querying it twice (which acquires a lock)
+            CancellationTokenSource tokenSource = InternalTokenSource;
+            if (tokenSource.IsCancellationRequested)
+            {
+                ThrowHelper.ThrowOperationCanceledException_ReadCanceled();
+            }
+
+            CancellationTokenRegistration reg = default;
+            if (cancellationToken.CanBeCanceled)
+            {
+                reg = cancellationToken.UnsafeRegister(state => ((StreamPipeReader)state!).Cancel(), this);
+            }
+
+            using (reg)
+            {
+                try
+                {
+                    if (_bufferedBytes > 0)
+                    {
+                        BufferSegment? segment = _readHead;
+                        try
+                        {
+                            while (segment != null)
```

Comment:
Iterate until _readTail, not null.

Reply:
Please see https://github.com/dotnet/runtime/pull/52159#discussion_r624684881